### ISSUE TITLE
Refactor io.py into a module

### ISSUE THREAD

**Issue by devin-petersohn** [modin-project/modin#203]
### Describe the problem

`modin.pandas.io` is currently a bit messy and needs a refactor. Ideally, we would refactor it into a module with python files for each backend (currently only Pandas on Ray).

This issue would be good for a first time contributor, so please feel free to reply to this thread with any questions if you are interested.

### Source code / logs

[Link to source](https://github.com/modin-project/modin/blob/master/modin/pandas/io.py)

**Comment by devin-petersohn** [comment]
I think we want a generally useful data ingest component to Modin. We don't necessarily want it to be tied to any specific API, so we will likely want to add some of the `io.py` functions to `modin.data_management`. Here is what I was thinking along that route:

We already use the factories to determine the runtime and format that Modin will use for objects that come from the pandas constructor (see [factories.py](https://github.com/modin-project/modin/blob/master/modin/data_management/factories.py)), but we don't have a way to do this with io. Ideally we would use the Factory classes to also serve data ingest (i.e. a classmethod to read from a given filetype. Then we would just use the same approach we do in `from_pandas`: `BaseFactory.read_csv` or something like that.

Right now, our parallel `read_csv` and `read_parquet` are the only two that would generally be affected by this. We should probably do the refactor in parts, but eventually all of the io should probably go directly through the factories. We can start with `read_csv` to see what makes sense.

We want to keep the error checking in `modin.pandas.io`, but move the Ray-specific implementations to the general `data_management` as a new module. Then we just need to add a classmethod similar to how we do `from_pandas`.

Let me know if you have any questions or if there's something unclear. Thanks for helping with this issue!

**Comment by eavidan** [comment]
Hey @devin-petersohn,
Started looking into this refactoring. 
I assume we will start from the same API call at `io.py`: `read_parquet()` and still move to `_read_parquet_pandas_on_ray()` with a small name change (without _on_ray).
There we can also perform our column names extraction from the file.

If I understand you correctly, we now return:
`return DataFrame(query_compiler=BaseFactory.read_parquet(path, engine, columns, **kwargs))`
following the flow of `from_pandas`, we will add the relevant methods to the BaseFactory and then the implementation to `PandasQueryCompiler`:
```
@classmethod
    def read_parquet(cls, path, engine, columns, block_partitions_cls, **kwargs):
```
Here we will compute the columns and send the work to Ray remote functions, using the block_partitions_cls.

The question is, where would we place the remote function at? and where should we invoke them from?
it feels like `PandasOnRayRemotePartition` should be part of this.
So, can you shed some light?
thanks

**Comment by devin-petersohn** [comment]
Hey @eavidan thanks for picking up this issue.

Your intuition is correct, but I think we want to create an `io` module in `modin.data_management`. The remote task should go in `modin.data_management.io.pandas_on_ray`. 

Following the hierarchy I believe this would be the most straightforward:

- `BaseFactory.read_parquet(**kwargs)`
    - `PandasOnRayFactory.read_parquet(**kwargs)`
        - `return modin.data_management.io.pandas_on_ray.read_parquet(**kwargs)`

Structurally this means you can probably just add a new field to `PandasOnRayFactory`:

```python
class PandasOnRayFactory(BaseFactory):
    ...
    io_module = modin.data_management.io.pandas_on_ray
```

From there the `BaseFactory` can just call the `read_parquet` of the `cls.io_module`. This will require a `_read_parquet` (similar to `from_pandas`).

```python
class BaseFactory(object):
    ...
    @classmethod
    def read_parquet(cls, **kwargs):
        return cls._determine_engine()._read_parquet(**kwargs)

    @classmethod
    def _read_parquet(cls, **kwargs):
        return cls.io_module.read_parquet(**kwargs)
```

The `*_pandas_on_ray` functions and remote tasks can just be put directly in the new file. Then you'll need a wrapper.

Please let me know if anything is unclear or if you have any concerns with the design. Thanks again for picking it up!

**Comment by devin-petersohn** [comment]
@eavidan I have implemented `read_table` in a recent PR (#270). This shouldn't change much on this refactor, but I am happy to help here too, especially if there are conflicts.

**Comment by eavidan** [comment]
This is the flow of things so far

- entry point is still in modin.pands.io, but looks like this
```py
def read_parquet(path, engine="auto", columns=None, **kwargs):
    """Load a parquet object from the file path, returning a DataFrame.

    Args:
        path: The filepath of the parquet file.
              We only support local files for now.
        engine: This argument doesn't do anything for now.
        kwargs: Pass into parquet's read_pandas function.
    """
    from ..data_management.factories import BaseFactory
    return DataFrame(query_compiler=BaseFactory.read_parquet(path=path, columns=columns, engine=engine, **kwargs))
```

- factories.py as you suggested
```py
    @classmethod
    def read_parquet(cls, **kwargs):
        return cls._determine_engine()._read_parquet(**kwargs)

    @classmethod
    def _read_parquet(cls, **kwargs):
        return cls.io_module.read_parquet(**kwargs)
```

- the `io_module` is now `modin.data_management.io.pands_on_ray.py, and the function in `read_parquet()` looks like this
```py
def read_parquet(**kwargs):
    """Load a parquet object from the file path, returning a DataFrame.
       Ray DataFrame only supports pyarrow engine for now.

    Args:
        path: The filepath of the parquet file.
              We only support local files for now.
        engine: Ray only support pyarrow reader.
                This argument doesn't do anything for now.
        kwargs: Pass into parquet's read_pandas function.

    Notes:
        ParquetFile API is used. Please refer to the documentation here
        https://arrow.apache.org/docs/python/parquet.html
    """
    return _read_parquet_pandas_on_ray(**kwargs)
```
`_read_parquet_pandas_on_ray` has the same implementation we had in the original `io.py`, but returns the `PandasQueryCompiler` instead of the DF.

@devin-petersohn could you please review the [full commit](https://github.com/eavidan/modin/commit/d2a15db905e8e267da044e64ea972eef3b19dcb8) to see we are aligned?
before I PR

**Comment by devin-petersohn** [comment]
Yes, I left a couple of minor comments on the commit, but that looks good. I think the structure looks much more general in the sense of supporting various backends from the pandas API. Thanks!

**Comment by devin-petersohn** [comment]
I believe this can be closed via #283, and #290
